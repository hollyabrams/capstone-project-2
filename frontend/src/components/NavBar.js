import Logo from "./Logo";
import { useState, useContext } from 'react';
import { NavLink } from 'react-router-dom';
import UserContext from '../UserContext';
import { ShoppingCartIcon } from '@heroicons/react/24/solid';
import { useShoppingCart } from "use-shopping-cart";

function ShoppingCart() {
    const { formattedTotalPrice, cartCount } = useShoppingCart();

    return (
        <NavLink to='/cart' className="flex items-center space-x-1 text-gray-700 hover:text-gray-900">
            <div className="relative">
                <ShoppingCartIcon className="w-7 h-7 flex-shrink-0"/>
            </div>
            <p className="text-lg">
                {formattedTotalPrice}
                <span className="text-sm text-gray-500" >({cartCount})</span>
            </p>
        </NavLink>
    );
}

function Navigation({ isOpen, loggedIn, logout, currentUser }) {
  const navItems = loggedIn
      ? [
          { to: "/products", name: "Products" },
          { to: "/profile", name: "Profile" },
          { to: "/", name: `Log out ${currentUser.first_name || currentUser.username}`, onClick: logout }
      ]
      : [
          { to: "/products", name: "Products" },
          { to: "/login", name: "Login" },
          { to: "/signup", name: "Sign Up" }
      ];

  return (
      <ul className={`flex items-center space-x-1 text-gray-700 hover:text-gray-900 ${isOpen ? '' : 'hidden'} sm:flex sm:items-center sm:justify-center sm:flex-grow`}>
          {navItems.map((item, idx) => (
              <li className="nav-item mx-4" key={idx}>
                  <NavLink className={`nav-link ${item.name === 'Login' || item.name === 'Profile' ? 'mr-4' : ''}`} to={item.to} onClick={item.onClick}>
                      {item.name}
                  </NavLink>
              </li>
          ))}
          <li className="sm:hidden">
              <ShoppingCart />
          </li>
      </ul>
  );
}

export default function NavBar() {
  const { currentUser, logout } = useContext(UserContext);
  const [isOpen, setIsOpen] = useState(false); 

  const toggleMenu = () => setIsOpen(!isOpen);

  console.debug('NavBar', 'currentUser=', currentUser);

  return (
      <nav className="sticky top-0 bg-white z-10 shadow">
          <div className="container mx-auto p-6 flex flex-wrap justify-between items-center">
              <Logo to="/" />
              <div className="px-4 cursor-pointer sm:hidden ml-auto absolute right-6 top-6" onClick={toggleMenu}>
                  <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" className="w-6 h-6">
                      <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M4 6h16M4 12h16M4 18h16" />
                  </svg>
              </div>
              <div className="flex-grow sm:flex sm:items-center sm:justify-center">
                  <Navigation isOpen={isOpen} loggedIn={!!currentUser} logout={logout} currentUser={currentUser} />
              </div>
              <div className="hidden sm:block">
                  <ShoppingCart />
              </div>
          </div>
      </nav>
  );
}

