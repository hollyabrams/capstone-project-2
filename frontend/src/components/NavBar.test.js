// NavBar.test.js
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import UserContext from '../UserContext';
import { useShoppingCart } from "use-shopping-cart";
import NavBar from './NavBar';

jest.mock("use-shopping-cart", () => ({
  useShoppingCart: () => ({
    formattedTotalPrice: "Â£0.00",
    cartCount: 0
  })
}));

jest.mock('../UserContext');

test('NavBar renders without crashing', () => {
  const mockContext = {
    currentUser: { username: "Test" },
    logout: jest.fn()
  };

  UserContext._currentValue = mockContext;
  
  render(
    <Router>
      <NavBar />
    </Router>
  );
  
  expect(screen.getByText(/Test/i)).toBeInTheDocument();
});

test('NavBar matches snapshot', () => {
  const mockContext = {
    currentUser: { username: "Test" },
    logout: jest.fn()
  };

  UserContext._currentValue = mockContext;

  const { asFragment } = render(
    <Router>
      <NavBar />
    </Router>
  );
  
  expect(asFragment()).toMatchSnapshot();
});
