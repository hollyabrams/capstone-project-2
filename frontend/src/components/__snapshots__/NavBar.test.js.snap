// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="sticky top-0 bg-white z-10 shadow"
  >
    <div
      class="container mx-auto p-6 flex flex-wrap justify-between items-center"
    >
      <a
        class="flex items-center space-x"
        href="/"
      >
        <img
          alt="logo"
          height="100"
          src="/logo.png"
          width="200"
        />
      </a>
      <div
        class="px-4 cursor-pointer sm:hidden ml-auto absolute right-6 top-6"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6h16M4 12h16M4 18h16"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <div
        class="flex-grow sm:flex sm:items-center sm:justify-center"
      >
        <ul
          class="flex items-center space-x-1 text-gray-700 hover:text-gray-900 hidden sm:flex sm:items-center sm:justify-center sm:flex-grow"
        >
          <li
            class="nav-item mx-4"
          >
            <a
              class="nav-link "
              href="/products"
            >
              Products
            </a>
          </li>
          <li
            class="nav-item mx-4"
          >
            <a
              class="nav-link mr-4"
              href="/profile"
            >
              Profile
            </a>
          </li>
          <li
            class="nav-item mx-4"
          >
            <a
              aria-current="page"
              class="nav-link  active"
              href="/"
            >
              Log out Test
            </a>
          </li>
          <li
            class="sm:hidden"
          >
            <a
              class="flex items-center space-x-1 text-gray-700 hover:text-gray-900"
              href="/cart"
            >
              <div
                class="relative"
              >
                <svg
                  aria-hidden="true"
                  class="w-7 h-7 flex-shrink-0"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
                  />
                </svg>
              </div>
              <p
                class="text-lg"
              >
                £0.00
                <span
                  class="text-sm text-gray-500"
                >
                  (0)
                </span>
              </p>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="hidden sm:block"
      >
        <a
          class="flex items-center space-x-1 text-gray-700 hover:text-gray-900"
          href="/cart"
        >
          <div
            class="relative"
          >
            <svg
              aria-hidden="true"
              class="w-7 h-7 flex-shrink-0"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
              />
            </svg>
          </div>
          <p
            class="text-lg"
          >
            £0.00
            <span
              class="text-sm text-gray-500"
            >
              (0)
            </span>
          </p>
        </a>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
